src/shared/edit-util.c

```diff
@@ -0,0 +1,378 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <stdio.h>
+
+#include "alloc-util.h"
+#include "copy.h"
+#include "edit-util.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "fs-util.h"
+#include "mkdir-label.h"
+#include "path-util.h"
+#include "process-util.h"
+#include "selinux-util.h"
+#include "stat-util.h"
+#include "string-util.h"
+#include "strv.h"
+#include "tmpfile-util.h"
+
+size_t edit_files_count(const EditFile *ef) {
+        size_t n = 0;
+
+        EDIT_FILE_FOREACH(i, ef)
+                n++;
+
+        return n;
+}
+
+void edit_file_free_all(EditFile **ef) {
+        assert(ef);
+
+        EDIT_FILE_FOREACH(i, *ef) {
+                free(i->path);
+                free(i->original_path);
+                free(i->temp);
+                strv_free(i->comment_paths);
+        }
+
+        *ef = mfree(*ef);
+}
+
+int create_edit_temp_file(
+                const char *target_path,
+                const char *original_path,
+                char * const *comment_paths,
+                const char *marker_start,
+                const char *marker_end,
+                char **ret_temp_filename,
+                unsigned *ret_edit_line) {
+
+        _cleanup_free_ char *temp = NULL;
+        unsigned line = 1;
+        int r;
+
+        assert(target_path);
+        assert(!comment_paths || (comment_paths && (marker_start || marker_end)));
+        assert(ret_temp_filename);
+
+        r = tempfn_random(target_path, NULL, &temp);
+        if (r < 0)
+                return log_error_errno(r, "Failed to determine temporary filename for \"%s\": %m", target_path);
+
+        r = mkdir_parents_label(target_path, 0755);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create parent directories for \"%s\": %m", target_path);
+
+        if (original_path) {
+                r = mac_selinux_create_file_prepare(target_path, S_IFREG);
+                if (r < 0)
+                        return r;
+
+                r = copy_file(original_path, temp, 0, 0644, 0, 0, COPY_REFLINK);
+                if (r == -ENOENT) {
+                        r = touch(temp);
+                        mac_selinux_create_file_clear();
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to create temporary file \"%s\": %m", temp);
+                } else {
+                        mac_selinux_create_file_clear();
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to create temporary file for \"%s\": %m", target_path);
+                }
+        }
+
+        if (comment_paths) {
```

Comment:
Previously, this was `else if`. Is this intended?

Reply:
Yes. I think the two don't really conflict, and can be used together if necessary.

Reply:
Maybe so, but I am not sure. Please do that change in a separate commit with explanation.